bitcoin/src/crypto/key.rs

```diff
@@ -863,76 +861,145 @@ impl From<TweakedKeypair> for TweakedPublicKey {
-    /// A secp256k1 error.
-    Secp256k1(secp256k1::Error),
+pub enum FromSliceError {
     /// Invalid key prefix error.
     InvalidKeyPrefix(u8),
-    /// Hex decoding error.
-    Hex(hex::HexToArrayError),
-    /// `PublicKey` hex should be 66 or 130 digits long.
-    InvalidHexLength(usize),
+    /// A Secp256k1 error.
+    Secp256k1(secp256k1::Error),
+    /// Invalid Length of the slice.
+    InvalidLength(usize),
 }
 
-impl fmt::Display for Error {
+impl fmt::Display for FromSliceError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        use Error::*;
+        use FromSliceError::*;
 
         match *self {
-            Base58(ref e) => write_err!(f, "base58"; e),
             Secp256k1(ref e) => write_err!(f, "secp256k1"; e),
             InvalidKeyPrefix(ref b) => write!(f, "key prefix invalid: {}", b),
-            Hex(ref e) => write_err!(f, "hex"; e),
-            InvalidHexLength(got) =>
-                write!(f, "pubkey hex should be 66 or 130 digits long, got: {}", got),
+            InvalidLength(got) => write!(f, "slice length should be 33 or 65 bytes, got: {}", got),
         }
     }
 }
 
 #[cfg(feature = "std")]
-impl std::error::Error for Error {
+impl std::error::Error for FromSliceError {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-        use Error::*;
+        use FromSliceError::*;
 
         match *self {
-            Base58(ref e) => Some(e),
             Secp256k1(ref e) => Some(e),
-            Hex(ref e) => Some(e),
-            InvalidKeyPrefix(_) | InvalidHexLength(_) => None,
+            InvalidKeyPrefix(_) | InvalidLength(_) => None,
         }
     }
 }
 
-impl From<base58::Error> for Error {
-    fn from(e: base58::Error) -> Error { Error::Base58(e) }
+impl From<secp256k1::Error> for FromSliceError {
+    fn from(e: secp256k1::Error) -> FromSliceError { Self::Secp256k1(e) }
+}
+
+/// Error generated from WIF key format.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum FromWifError {
+    /// A base58 error.
+    Base58(base58::Error),
+    /// A secp256k1 error.
+    Secp256k1(secp256k1::Error),
+}
+
+impl fmt::Display for FromWifError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use FromWifError::*;
+        match *self {
+            Base58(ref e) => write_err!(f, "invalid base58"; e),
+            Secp256k1(ref e) => write_err!(f, "private key validation failed"; e),
+        }
+    }
+}
+
+impl From<base58::Error> for FromWifError {
+    fn from(e: base58::Error) -> FromWifError { Self::Base58(e) }
+}
+
+impl From<secp256k1::Error> for FromWifError {
+    fn from(e: secp256k1::Error) -> FromWifError { Self::Secp256k1(e) }
+}
+
+/// Error returned while constructing public key from string.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ParsePublicKeyError {
+    /// Error originated while parsing string.
+    Encoding(FromSliceError),
+    /// Hex decoding error.
+    Hex(hex::HexToArrayError),
```

Comment:
This error contains invalid length variant which is also present in `FromSliceError`. This sucks because it's confusing and not clear that the invalid length in `HexToArray` is actually unreachable (because we pre-check it ina a `match`).

Side note: also this leaks hex in API but that's for a future PR.

I propose that we have `InvalidChar` (because odd length is also unreachable) here directly and then just panic on unreachable variants.

Reply:
> This error contains invalid length variant which is also present in `FromSliceError`. This sucks because it's confusing and not clear that the invalid length in `HexToArray` is actually unreachable (because we pre-check it ina a `match`).

Actually if we remove the `HexToArray`. There could be an error while parsing the hex. How should we deal with that?

```rust
pub enum HexToArrayError {
    /// Conversion error while parsing hex string.
    Conversion(HexToBytesError),
    /// Tried to parse fixed-length hash from a string with the wrong length (got, want).
    InvalidLength(usize, usize),
}
```
> I propose that we have `InvalidChar` (because odd length is also unreachable) here directly and then just panic on unreachable variants.

Can you explain a bit more.

Reply:
Yes, but the only possible hex error is invalid char because we check the length upfront.

So just use `InvalidChar(u8)` in place of `Hex(HexToArrayError)` and match on the result when converting, panicking on other variants.

(And FTR it'd be much better to have an `InvalidCharError` but that's an issue for the `he-conservative` crate.)

Reply:
Got it. Only the `InvalidChar` can occur in this case. Also the `OddLengthString` is also pre-checked. Thanks!